invoice/invoice.go

```diff
@@ -0,0 +1,966 @@
+// be implied.
+func Expiry(expiry uint64) func(*Invoice) {
+	return func(i *Invoice) {
+		i.Expiry = &expiry
+	}
+}
+
+// FallbackAddr is a functional option that allows callers of NewInvoice to set
+// the Invoice's fallback on-chain address that can be used for payment in case
+// the Lightning payment fails
+func FallbackAddr(fallbackAddr btcutil.Address) func(*Invoice) {
+	return func(i *Invoice) {
+		i.FallbackAddr = fallbackAddr
+	}
+}
+
+// RoutingInfo is a functional option that allows callers of NewInvoice to set
+// one or more entries containing extra routing information for a private route
+// to the target node.
+func RoutingInfo(routingInfo []ExtraRoutingInfo) func(*Invoice) {
+	return func(i *Invoice) {
+		i.RoutingInfo = routingInfo
+	}
+}
+
+// NewInvoice creates a new Invoice object. The last parameter is a set of
+// variadic argumements for setting optional fields of the invoice.
+// Note: Either Description  or DescriptionHash must be provided for the Invoice
+// to be considered valid.
+func NewInvoice(net *chaincfg.Params, paymentHash []byte,
+	timestamp uint64, options ...func(*Invoice)) *Invoice {
+
+	invoice := &Invoice{
+		Net:         net,
+		PaymentHash: paymentHash,
+		Timestamp:   timestamp,
+	}
+
+	for _, option := range options {
+		option(invoice)
+	}
+
+	return invoice
+}
+
+// Decode parses the provided encoded invoice, and returns a decoded Invoice in
+// case it is valid by BOLT-0011.
+func Decode(invoice string) (*Invoice, error) {
+	decodedInvoice := Invoice{}
+
+	// Decode the invoice using the modified bech32 decoder.
+	hrp, data, err := DecodeBech32(invoice)
+	if err != nil {
+		return nil, err
+	}
+
+	// We expect the human-readable part to at least have ln + two chars
+	// encoding the network.
+	if len(hrp) < 4 {
+		return nil, fmt.Errorf("hrp too short")
+	}
+
+	// First two characters of HRP should be "ln".
+	if hrp[:2] != "ln" {
+		return nil, fmt.Errorf("prefix should be \"ln\"")
+	}
+
+	// The next two characters should be a valid prefix for a segwit BIP173
+	// address.
+	segwitPrefix := hrp[2:4] + "1"
+	if !chaincfg.IsBech32SegwitPrefix(segwitPrefix) {
+		return nil, fmt.Errorf("unknown network \"%v\"", hrp[2:4])
+	}
+
+	// Extract which network this invoice is meant for.
+	var net *chaincfg.Params
+	switch hrp[2:4] {
+	case "bc":
+		net = &chaincfg.MainNetParams
+	case "tb":
+		net = &chaincfg.TestNet3Params
+	case "sb":
+		net = &chaincfg.SimNetParams
+	default:
+		return nil, fmt.Errorf("unknown network: %v", hrp[2:4])
+	}
+	decodedInvoice.Net = net
+
+	// Optionally, if there's anything left of the HRP, it encodes the
+	// payment amount.
+	if len(hrp) > 4 {
+		amount, err := decodeAmount(hrp[4:])
+		if err != nil {
+			return nil, err
+		}
+		decodedInvoice.MilliSat = &amount
+	}
+
+	// Everything except the last 520 bits of the data encodes the invoice's
+	// timestamp and tagged fields.
+	invoiceData := data[:len(data)-104]
```

Comment:
The `104` value should be lifted into a constant above with an appropriate `godoc` comment explaining that since `bech32` works on 5-bit words we extract 104 words, so 520 bytes.